src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedInlineArrayTypeSymbol.cs

```diff
@@ -0,0 +1,245 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.CSharp.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// A synthesized type used during emit to allow temp locals of Span&lt;T&gt;
+    /// of a specific length where the span storage is on the stack.
+    /// <code>
+    /// [InlineArray(N)] struct __InlineArrayN&lt;T&gt; { private T _element0; }
+    /// </code>
+    /// </summary>
+    internal sealed class SynthesizedInlineArrayTypeSymbol : NamedTypeSymbol
+    {
+        private readonly ModuleSymbol _containingModule;
+        private readonly int _arrayLength;
+        private readonly MethodSymbol _inlineArrayAttributeConstructor;
+        private readonly ImmutableArray<FieldSymbol> _fields;
+
+        internal SynthesizedInlineArrayTypeSymbol(SourceModuleSymbol containingModule, string name, int arrayLength, MethodSymbol inlineArrayAttributeConstructor)
+        {
+            Debug.Assert(arrayLength > 0);
+
+            var typeParameter = new InlineArrayTypeParameterSymbol(this);
+            var field = new SynthesizedFieldSymbol(this, typeParameter, "_element0");
+
+            _containingModule = containingModule;
+            _arrayLength = arrayLength;
+            _inlineArrayAttributeConstructor = inlineArrayAttributeConstructor;
+            _fields = ImmutableArray.Create<FieldSymbol>(field);
+            Name = name;
+            TypeParameters = ImmutableArray.Create<TypeParameterSymbol>(typeParameter);
+        }
+
+        public override int Arity => 1;
+
+        public override ImmutableArray<TypeParameterSymbol> TypeParameters { get; }
+
+        public override NamedTypeSymbol ConstructedFrom => this;
+
+        public override bool MightContainExtensionMethods => false;
+
+        public override string Name { get; }
+
+        public override IEnumerable<string> MemberNames => GetMembers().SelectAsArray(m => m.Name);
+
+        public override Accessibility DeclaredAccessibility => Accessibility.Internal;
+
+        public override bool IsSerializable => false;
+
+        public override bool AreLocalsZeroed => true;
+
+        public override TypeKind TypeKind => TypeKind.Struct;
+
+        public override bool IsRefLikeType => false;
+
+        public override bool IsReadOnly => true;
+
+        public override Symbol? ContainingSymbol => _containingModule.GlobalNamespace;
```

Comment:
Is this the right containing symbol? I thought this is a nested type inside the private implementation type (based on `_orderedNestedTypes = orderedProxyTypes.Concat(orderedInlineArrayTypes)`)

Reply:
The types are actually top-level types even though they are visited in emit through `PrivateImplementationDetails.GetNestedTypes()`.

Reply:
Let's test that they can be referenced in any way in source (when loading from metadata), not merely inaccessible.

Reply:
From offline discussion, let's leave a comment here if the symbol cannot reference the proper containing symbol. Let's try to emit those types are nested types in the private implementation.

Reply:
Changed inline array types to be top-level types with names that cannot be used in source.